### ISSUE TITLE
[2.0] Keep high-level API as similar to 1.x as possible

### ISSUE THREAD

**Issue by colinodell** [thephpleague/commonmark#475]
Many users only care about adding extensions, setting configs, and converting Markdown.  Let's try to keep those APIs as similar as possible, even if a lot of the "internal" bits have BC-breaks.

This will require a review of those APIs and possibly a new 1.x minor release to denote deprecations to end-users.

(For example, we could add class aliases in 1.x so that users can be compatible with both that newer 1.x release and 2.x.  This may not be possible or feasible in all cases though - needs more research)

A sample, incomplete list of such things:

 - `CommonMarkConverter`
 - `ExtensionInterface`
 - Things in the new `Environment` sub-namespace

**Comment by colinodell** [comment]
Let's try to get as much forward-compatibility and deprecation notices into 1.5 as possible :wink:

**Comment by colinodell** [comment]
# Current status (WIP)

(This will be updated as I review the classes over the comings days and weeks)

## ✔ CommonMarkConverter

In 2.x, the base `Converter` class is gone, but all the other important bits should be compatible:

| Feature | Similar Enough? | Notes |
| --- | --- | --- |
| Base interface | ✔ | Both implement `MarkdownConverterInterface` |
| `__construct()` | ✔ | The `EnvironmentInterface` is in a different namespace, but the argument position is the same |
| `getEnvironment()` | ✔ | The `EnvironmentInterface` is in a different namespace |
| `convertToHtml()` | ✔ | Identical |
| `__invoke()` | ✔ | Identical |
| `public const VERSION` | ✔ | Identical  |

## ConfigurableEnvironmentInterface

A couple of methods have changed, and this lives in a different namespace now, but the method to add extensions is the same:

| Feature | Similar Enough? | Notes |
| --- | --- | --- |
| `addExtension()` | ✔ | Identical |

**TODO:** Update this based on #436 and #447

**Comment by colinodell** [comment]
**TL;DR:** After discussing with Mark Carver I've decided that, in the best interest of this project and its users, no further effort will be put into forward-compatibility for 1.5; instead, we'll focus on minimizing BC-breaks in 2.x and making that version the best it can be.  See the longer thread below for why.

-----

Generally speaking, there are three types of people who use this library:

1. People who instantiate a `CommonMarkConverter`, maybe add some extensions, and convert Markdown
2. People who integrate this library with other projects/frameworks and who use some of the other high-level APIs, like extensions
3. People who create custom parsers, event listeners, etc.

When considering the potential BC-breaks between 1.x and 2.0, the first group of users will be largely unaffected, as those highest-level APIs are remaining consistent enough that there will be no noticeable impact.

The third group will experience multiple BC breaks which are unfortunately unavoidable - we couldn't implement a full forward-compatibility layer in 1.5 even if we wanted to.

That leaves the second group of people.

Originally I was hoping to use [the class_alias trick](https://dev.to/greg0ire/how-to-deprecate-a-type-in-php-48cf) to provide a "forward-compatibility" layer so that this group could set their constraints to `^1.5 | ^2.0` and leverage the new high-level APIs coming in 2.x.  I even put together a proof-of-concept which Mark Carver helped to test, and the results were very promising - it technically worked perfectly, but with a few drawbacks:

1. PhpStorm didn't understand the voodoo magic going on and would display error messages for things that wouldn't actually produce errors
2. It would take many hours to properly implement and test this, taking my focus away from everything else going on
3. A single small mistake in this layer would produce fatal errors
4. **It would only be useful for people who use a `^1.5 | ^2.0` constraint** - anyone wanting to support old versions couldn't leverage it, and anyone supporting `^2.0` only wouldn't need it

Based on these factors, and the very productive conversation Mark and I had about this via Slack today, we've decided that it doesn't make sense to make 1.5 as forward-compatible as possible - the amount of work and risk is too high compared to the potential benefits.

Therefore, we will instead be flagging all deprecations in the code and CHANGELOG as usual.  **No special forward-compatibility layer will be added to 1.5**.

I will be keeping this issue open however, but switching the milestone to 2.0.  **Instead of looking for opportunities to create forward-compatibility, we'll instead look at maintaining backward-compatibility.**  In other words, we'll re-review the changes coming in 2.0 and see if there are any opportunities to keep things like method signatures as close as possible, and then **we'll document all the important things that are staying the same** to help people better understand the upgrade path.

As much as I hate breaking BC, 2.0 is doing it for very good reasons, and I strongly believe that users would benefit **more** from a well-written 2.x series that is built to last, rather than from us putting too much effort into something only a handful of people might use.

-----

Please know that regardless of this decision, this project remains 100% dedicated to:

1. Following semver and [BC best practices](https://symfony.com/doc/current/contributing/code/bc.html)
2. Minimizing BC-breaks in major versions as much as possible
3. Making it as simple as possible for most users to upgrade
4. Providing a strong architecture in 2.x that will last as long as possible
5. Making the experience of working with this library the best that it can be :)
